### PR TITLE
EngineNewPayloadV2  -  Reject non-null Blob fields 

### PR DESCRIPTION
**Pull request by Gabriel-Trintinalia**
## PR description
Validate blobGasUsed and excessblobGas pre-Cancun.

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2.java [hunk 1]
```diff
@@ -48,6 +49,22 @@ public String getName() {
...
   }
 
+  @Override
+  protected ValidationResult<RpcErrorType> validateParameters(
+      final EnginePayloadParameter payloadParameter,
+      final Optional<List<String>> maybeVersionedHashParam,
+      final Optional<String> maybeBeaconBlockRootParam) {
+    if (payloadParameter.getBlobGasUsed() != null) {
+      return ValidationResult.invalid(
+          RpcErrorType.INVALID_PARAMS, "non-nil BlobGasUsed pre-cancun");
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2.java:59]:
```suggestion
          RpcErrorType.INVALID_PARAMS, "non-null BlobGasUsed pre-cancun");
```

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2.java [hunk 2]
```diff
@@ -48,6 +49,22 @@ public String getName() {
...
+    }
+    if (payloadParameter.getExcessBlobGas() != null) {
+      return ValidationResult.invalid(
+          RpcErrorType.INVALID_PARAMS, "non-nil ExcessBlobGas pre-cancun");
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2.java:63]:
```suggestion
          RpcErrorType.INVALID_PARAMS, "non-null ExcessBlobGas pre-cancun");
```

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2Test.java [hunk 3]
```diff
@@ -127,6 +128,37 @@ public void shouldReturnInvalidIfWithdrawalsIsNotNull_WhenWithdrawalsProhibited(
...
+    BlockHeader blockHeader =
+        createBlockHeaderFixture(Optional.of(Collections.emptyList()), Optional.empty())
+            .excessBlobGas(BlobGas.MAX_BLOB_GAS)
+            .buildHeader();
+
+    var resp = resp(mockEnginePayload(blockHeader, Collections.emptyList(), List.of(), null));
+
+    final JsonRpcError jsonRpcError = fromErrorResp(resp);
+    assertThat(jsonRpcError.getCode()).isEqualTo(INVALID_PARAMS.getCode());
+    assertThat(jsonRpcError.getData()).isEqualTo("non-nil ExcessBlobGas pre-cancun");
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2Test.java:158]:
```suggestion
    assertThat(jsonRpcError.getData()).isEqualTo("non-null ExcessBlobGas pre-cancun");
```

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2Test.java [hunk 4]
```diff
@@ -127,6 +128,37 @@ public void shouldReturnInvalidIfWithdrawalsIsNotNull_WhenWithdrawalsProhibited(
...
+    // V2 should return error if non-null blobGasUsed
+    BlockHeader blockHeader =
+        createBlockHeaderFixture(Optional.of(Collections.emptyList()), Optional.empty())
+            .blobGasUsed(100L)
+            .buildHeader();
+
+    var resp = resp(mockEnginePayload(blockHeader, Collections.emptyList(), List.of(), null));
+    final JsonRpcError jsonRpcError = fromErrorResp(resp);
+    assertThat(jsonRpcError.getCode()).isEqualTo(INVALID_PARAMS.getCode());
+    assertThat(jsonRpcError.getData()).isEqualTo("non-nil BlobGasUsed pre-cancun");
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2Test.java:142]:
```suggestion
    assertThat(jsonRpcError.getData()).isEqualTo("non-null BlobGasUsed pre-cancun");
```

**Review by macfarla — APPROVED** [review]
nits re nil vs null, otherwise LGTM

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2Test.java [hunk 5]
```diff
@@ -127,6 +128,37 @@ public void shouldReturnInvalidIfWithdrawalsIsNotNull_WhenWithdrawalsProhibited(
...
+            .buildHeader();
+
+    var resp = resp(mockEnginePayload(blockHeader, Collections.emptyList(), List.of(), null));
+
+    final JsonRpcError jsonRpcError = fromErrorResp(resp);
+    assertThat(jsonRpcError.getCode()).isEqualTo(INVALID_PARAMS.getCode());
+    assertThat(jsonRpcError.getData()).isEqualTo("non-null ExcessBlobGas pre-cancun");
+    verify(engineCallListener, times(1)).executionEngineCalled();
+  }
+
```
Comment (pinges) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2Test.java:161]: Would be nice to have a test for parentBeaconBlockRoot as well!

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2.java [hunk 6]
```diff
@@ -48,6 +49,22 @@ public String getName() {
...
+      final Optional<List<String>> maybeVersionedHashParam,
+      final Optional<String> maybeBeaconBlockRootParam) {
+    if (payloadParameter.getBlobGasUsed() != null) {
+      return ValidationResult.invalid(
+          RpcErrorType.INVALID_PARAMS, "non-null BlobGasUsed pre-cancun");
+    }
+    if (payloadParameter.getExcessBlobGas() != null) {
+      return ValidationResult.invalid(
+          RpcErrorType.INVALID_PARAMS, "non-null ExcessBlobGas pre-cancun");
+    }
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV2.java:64]: parentBeaconBlockRoot schould be tested here as well

**Review by pinges — COMMENTED** [review]
We should add the parent beacon block root in here as well

**Pull request merged**